### PR TITLE
Add DotFolderByPath GraphQL root query for folder tree browsing

### PR DESCRIPTION
**Pull request by fmontes**
https://github.com/user-attachments/assets/962ab9d8-9eac-44cb-8dfd-4543105193f9

## Summary

Closes #34692

Adds a new `DotFolderByPath` GraphQL root query that returns a folder at a given path with recursive `children` traversal, enabling full folder tree browsing.

**Example:**
```graphql
query {
  DotFolderByPath(path: "/application/") {
    folderTitle
    folderPath
    children {
      folderTitle
      folderPath
    }
  }
}
```

### Implementation

Follows the established provider pattern (same as `DotNavigation`):

- **`FolderCollectionTypeProvider`** — defines `DotFolderCollectionItem` type with recursive `children` field using `GraphQLTypeReference`
- **`FolderCollectionFieldProvider`** — defines the `DotFolderByPath(path: String!)` root query field returning a single folder object
- **`FolderCollectionDataFetcher`** — resolves the folder at the given path via `FolderAPI.findFolderByPath()`, eagerly loads children with configurable max depth (`GRAPHQL_FOLDER_COLLECTION_MAX_DEPTH`, default 3)
- **`GraphqlAPIImpl`** — registered both new providers

### Design decisions

- **Query name `DotFolderByPath`** instead of `DotFolder` to avoid conflict with the existing `DotFolder` type in `CustomFieldType.FOLDER`
- **Max depth default 3** (configurable via env var) to limit eager loading performance impact
- **Children return `null` on error** instead of empty list, so clients can distinguish "no children" from "error loading children"
- **Specific exception handling** — catches `DotSecurityException` (permission denied), `DotDataException` (data layer), and generic `Exception` separately with distinct log messages
- **`folderId` uses `folder.getInode()`** — consistent with existing `DotFolderTransformerImpl`

### Files changed

| File | Change |
|------|--------|
| `FolderCollectionTypeProvider.java` | New — recursive GraphQL type |
| `FolderCollectionFieldProvider.java` | New — root query field |
| `FolderCollectionDataFetcher.java` | New — data fetcher with depth guard |
| `FolderCollectionDataFetcherTest.java` | New — unit tests |
| `GraphqlAPIImpl.java` | Modified — register providers |

## Test plan

- [x] Unit tests for `FolderCollectionDataFetcher`
- [x] Build passes (`./mvnw install -pl :dotcms-core -DskipTests`)
- [ ] Manual test: query `DotFolderByPath(path: "/")` returns root folder with children
- [ ] Manual test: nested `children` queries resolve correctly
- [ ] Manual test: non-existent path returns `null`
- [ ] Integration tests verify schema structure

🤖 Generated with [Claude Code](https://claude.com/claude-code)

This PR fixes: #34692

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        final HttpServletRequest request = context.getHttpServletRequest();
+
+        final String path = environment.getArgument("path");
+
+        if (!UtilMethods.isSet(path)) {
+            return null;
+        }
+
+        final Host host = WebAPILocator.getHostWebAPI().getHost(request);
+
```
[the remark this answers is not in the document]
Comment (fmontes) [reply dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java:102]: Fixed in 5e1da320e5. Added null check for host before passing to `findFolderByPath` — returns null with a warn log if host cannot be resolved from the request.

File: dotCMS/src/main/java/com/dotcms/graphql/business/FolderCollectionTypeProvider.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    final Map<String, GraphQLOutputType> folderCollectionFields = createFolderCollectionFields();
+
+    final GraphQLObjectType folderCollectionType = createObjectType(
+            DOT_FOLDER_COLLECTION_ITEM, folderCollectionFields, null);
+
+    @Override
+    public Collection<? extends GraphQLType> getTypes() {
+        Logger.debug(this, () -> "Creating DotFolderCollectionItem types");
```
[the remark this answers is not in the document]
Comment (fmontes) [reply dotCMS/src/main/java/com/dotcms/graphql/business/FolderCollectionTypeProvider.java]: Fixed in 5e1da320e5. Removed the space before the arrow to match `NavigationTypeProvider` style: `()->`"Creating DotFolderCollectionItem types"`.

File: dotCMS/src/main/java/com/dotcms/graphql/business/GraphqlAPIImpl.java [hunk 3]
```diff
@@ -57,12 +57,14 @@ protected GraphqlAPIImpl(final GraphQLSchemaCache schemaCache) {
         typesProviders.add(QueryMetadataTypeProvider.INSTANCE);
         typesProviders.add(PaginationTypeProvider.INSTANCE);
         typesProviders.add(NavigationTypeProvider.INSTANCE);
+        typesProviders.add(FolderCollectionTypeProvider.INSTANCE);
```
[the remark this answers is not in the document]
Comment (fmontes) [reply dotCMS/src/main/java/com/dotcms/graphql/business/GraphqlAPIImpl.java:60]: Not needed — `FolderCollectionTypeProvider` and `FolderCollectionFieldProvider` are in the same package (`com.dotcms.graphql.business`) as `GraphqlAPIImpl`. Java does not require imports for same-package classes.

File: dotCMS/src/main/java/com/dotcms/graphql/business/GraphqlAPIImpl.java [hunk 4]
```diff
@@ -57,12 +57,14 @@ protected GraphqlAPIImpl(final GraphQLSchemaCache schemaCache) {
...
         //Register Fields
         fieldsProviders.add(ContentAPIGraphQLFieldsProvider.INSTANCE);
         fieldsProviders.add(PageAPIGraphQLFieldsProvider.INSTANCE);
         fieldsProviders.add(QueryMetadataFieldProvider.INSTANCE);
         fieldsProviders.add(PaginationFieldProvider.INSTANCE);
         fieldsProviders.add(NavigationFieldProvider.INSTANCE);
+        fieldsProviders.add(FolderCollectionFieldProvider.INSTANCE);
```
[the remark this answers is not in the document]
Comment (fmontes) [reply dotCMS/src/main/java/com/dotcms/graphql/business/GraphqlAPIImpl.java:67]: Same as above — same package, no import needed.

File: dotCMS/src/test/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcherTest.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        folder.setInode(inode);
+        folder.setName(name);
+        folder.setPath(path);
+        folder.setTitle(title);
+        folder.setFilesMasks(filesMask);
+        folder.setSortOrder(sortOrder);
+        folder.setDefaultFileType(defaultFileType);
+        return folder;
+    }
+}
```
[the remark this answers is not in the document]
Comment (fmontes) [reply dotCMS/src/test/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcherTest.java:170]: Valid concern — integration tests for permission handling, FolderAPI interaction, and depth limiting would add real value. Deferring to a follow-up: unit tests for the current behavior are in place and the PR scope is focused on the feature itself.

File: dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        map.put("folderDefaultFileType", folder.getDefaultFileType());
+
+        if (depth >= maxDepth) {
+            Logger.warn(this, "Max folder depth (" + maxDepth
+                    + ") reached for folder: " + folder.getPath());
+            map.put("children", Collections.emptyList());
+            return map;
+        }
+
+        try {
```
Comment (hassandotcms) [inline dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java:200]: we are catching exceptions here, is it intentional to not inform caller about errors in this context?

File: dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        final User user = context.getUser();
+        final HttpServletRequest request = context.getHttpServletRequest();
+
+        final String path = environment.getArgument("path");
+
+        if (!UtilMethods.isSet(path)) {
+            return null;
+        }
+
+        final Host host = WebAPILocator.getHostWebAPI().getHost(request);
```
Comment (dsilvam) [inline dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java]: Should be also take the host from params (optional) and if not provided default to this?

**Review by fabrizzio-dotCMS — COMMENTED** [review]
A postman test is in order here

File: dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+                    .collect(Collectors.toList());
+            map.put("children", childMaps);
+        } catch (Exception e) {
+            Logger.error(this, "Error loading children for folder: "
+                    + folder.getPath(), e);
+            map.put("children", null);
+        }
+
+        return map;
+    }
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java:242]: I agree. It is important to test the recursion, especially the edge cases: subfolders with limited permissions, where an exception is thrown

File: dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+        map.put("folderId", folder.getInode());
+        map.put("folderFileMask", folder.getFilesMasks());
+        map.put("folderSortOrder", folder.getSortOrder());
+        map.put("folderName", folder.getName());
+        map.put("folderPath", folder.getPath());
+        map.put("folderTitle", folder.getTitle());
+        map.put("folderDefaultFileType", folder.getDefaultFileType());
+
+        if (depth >= maxDepth) {
+            Logger.warn(this, "Max folder depth (" + maxDepth
```
[the remark this answers is not in the document]
Comment (nollymar) [reply dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java]: Let's change the logging level here @fmontes . Debug will generate less noise

File: dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java (hunk 7 above)
Comment (fmontes) [reply dotCMS/src/main/java/com/dotcms/graphql/datafetcher/FolderCollectionDataFetcher.java]: Added.

**Review by dario-daza — APPROVED** [review]

**Review by dsilvam — APPROVED** [review]

**Review by hassandotcms — APPROVED** [review]

**Review by nollymar — APPROVED** [review]
